.github/workflows/regressions.yml

```diff
@@ -0,0 +1,116 @@
+jobs:
+    calculate-base-commit:
+        runs-on: ubuntu-latest
+        outputs:
+            base-commit: ${{ steps.base-commit.outputs.base-commit }}
+        steps:
+            - uses: actions/checkout@v2
+              with:
+                  fetch-depth: 0
+
+            # In case of `workflow_dispatch` event, `github.sha` points to latest commit of chosen branch
+            # So, we need to find the latest common commit for master branch and chosen one
+            # to calculate difference only for branch changes
+            - name: Calculate base commit for workflow_dispatch event
+              if: github.event_name == 'workflow_dispatch'
+              run: echo "::set-env name=BASE_COMMIT::$(git merge-base origin/master ${{ github.sha }})"
+
+            # For pull request event, GitHub produces additional merge commit with `master` branch and PR branch as parents
+            # In this case, we want to check difference between master branch and merge commit
+            # so emit hash of `origin/master` branch itself as base commit
+            - name: Calculate base commit for pull_request event
+              if: github.event_name == 'pull_request'
+              run: echo "::set-env name=BASE_COMMIT::$(git log origin/master -n 1 --format=format:%H)"
+
+            - name: Emit base commit
+              id: base-commit
+              run: echo "::set-output name=base-commit::$BASE_COMMIT"
+
+            - name: Show commits
+              run: |
+                  echo "current commit:"
+                  git log ${{ github.sha }} -n 1 --pretty=short
+                  echo "base commit:"
+                  git log ${{ steps.base-commit.outputs.base-commit }} -n 1 --pretty=short
+
+    check:
+        needs: [ calculate-base-commit ]
+        name: ${{ matrix.project.name }}
+        runs-on: ubuntu-latest
+        strategy:
+            fail-fast: false
+            matrix:
+                project:
+                    - name: cargo
+                      path: cargo
+                      repository: rust-lang/cargo
+                      exclude_paths: ""
+        env:
+            PROJECT_PATH: ${{ matrix.project.path }}
+            PROJECT_URL: https://github.com/${{ matrix.project.url }}
+            PROJECT_EXCLUDE_PATHS: ${{ matrix.project.exclude_paths }}
+            ORG_GRADLE_PROJECT_showStandardStreams: true
+        steps:
+            - uses: actions/checkout@v2
+              with:
+                  fetch-depth: 0
+
+            - uses: actions/checkout@v2
+              with:
+                  repository: ${{ matrix.project.repository }}
+                  ref: master
+                  path: testData/${{ matrix.project.name }}
+
+            - name: Set up JDK 1.8
+              uses: actions/setup-java@v1
+              with:
+                  java-version: 1.8
+
+            - name: Set up Python
+              uses: actions/setup-python@v1
+              with:
+                  python-version: 3.7
```

Comment:
I'd personally write this in Kotlin/Rust. Python is undoubtedly the right tool for the job here, but, imo, adding a new technology has high cost. Even if Kotlin is less prettier to run, I *think* it will be easier to maintain long-term.

Reply:
But, obviously, no strong opinion here :D

Reply:
We already used some other python scripts in other workflows so it's not new technology for project.
Unfortunately, there is an issue with jvm-based language here (even if we forget that gradle is insanely slow). If you try to run any kotlin script via gradle in this project, gradle downloads IDEA & CLion dependencies despite we don't run anything related to plugin itself because gradle intellij plugin still downloads IDE dependencies in configuration phase. The [corresponding](https://github.com/JetBrains/gradle-intellij-plugin/issues/123) bug is fixed only recently and is not released yet. So any gradle run leads to downloading about 2GB and it takes more than 2 minutes.
Of course, in this concrete case we need to download these dependencies anyway but it's quite weird to use different approaches in different CI tasks
In combination with slowness of gradle itself, necessity of compilation and small probability of complex python scripts (in my opinion), it makes run/debug workflow quite slow for me (even with fact that I don't know python enough).
Anyway, if we'll understand that maintaining become hard, we will rewrite these scripts

Reply:
Hm, that's a bummer indeed... And, apparently, there's no easy way to write Kotlin scripts without gradle? It'd be really nice if one could just `curl` a minimal Kotlin runtime with popular libraries, just for bashesque scripting. Might be a fun project for hackathone.

In rust-analyzer, we write all our bash [in Rust](https://github.com/rust-analyzer/rust-analyzer/tree/master/xtask), and it works so much better than gradle in practice :-) 

I guess, 

>The right tool for the job is often the tool you are already using —adding new tools has a higher cost than many people appreciate. 

has an additional condition -- existing tool is Rust :D